File: Algorithm-Repository/Basic-Arithmetic/Cpp/IntegerDivision.cpp
// IntegerDivision.cpp
// Language: C++
// Purpose: perform integer division, floor division, and ceiling division
// Time Complexity: O(1)
// Space Complexity: O(1)
// Use Case: Useful in array partitioning, paging, and problem-solving in competitive programming

#include <iostream>
using namespace std;

int main() {
    int a, b;
    cout << "enter dividend and divisor: ";
    cin >> a >> b;

    if(b == 0) {
        cout << "division by zero is not allowed!" << endl;
        return 1;
    }

    int iDiv = a / b;                 // this do integer division
    int fDiv = (a >= 0) ? a / b : (a - b + 1) / b;   // this do floor division
    int cDiv = (a + b - 1) / b;      // this do ceiling division

    cout << "integer Division (truncated): " << iDiv << endl;
    cout << "Floor Division: " << fDiv << endl;
    cout << "Ceiling Division: " << cDiv << endl;

    return 0;
}

File: Algorithm-Repository/Basic-Arithmetic/Java/IntegerDivision.java
// IntegerDivision.java
// Language: Java
// Purpose: perform integer division, floor division, and ceiling division
// Time Complexity: O(1)
// Space Complexity: O(1)
// Use Case: Useful in array partitioning, paging, and problem-solving in competitive programming

import java.util.Scanner;

public class IntegerDivision {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("enter dividend and divisor: ");
        int a = sc.nextInt();
        int b = sc.nextInt();

        if(b == 0) {
            System.out.println("division by zero is not allowed!");
            sc.close();
            return;
        }

        int iDiv = a / b;                        
        int fDiv = (a >= 0) ? a / b : (a - b + 1) / b;
        int cDiv = (a + b - 1) / b;            
        
        System.out.println("Integer division : " + iDiv);
        System.out.println("Floor division: " + fDiv);
        System.out.println("Ceiling division: " + cDiv);


        sc.close();
    }
}
